### PR TITLE
tarFilterer: handle multiple archives

### PR DESCRIPTION
**Pull request by nalind**
#### What type of PR is this?

/kind other

#### What this PR does / why we need it:

When processing `.dockerignore` files, we produce mini-archives for each individual item being copied as part of a directory tree, so if we pass them through a `tarFilterer`, it needs to not give up when it hits the end of the first archive that it sees.  Add a flag to it that it sets when it closes the writing end of the pipe that it's reading, and if the flag isn't set when the reader finds an EOF, have the reader gear up to read another archive instead of just refusing to process any more data.

#### How to verify it

New unit tests exercise the code path that would have been tripped up.

#### Which issue(s) this PR fixes:

`None`

#### Special notes for your reviewer:

This expands on #2479 by giving the callback that tar filterers use more options, again inspired by `imagebuilder`.

#### Does this PR introduce a user-facing change?

```
None
```

### CODE REVIEW

File: digester.go [hunk 1]
```diff
@@ -62,50 +63,65 @@ func (t *tarFilterer) Write(p []byte) (int, error) {
 }
 
 func (t *tarFilterer) Close() error {
+	t.closed = true
```
Comment (rhatdan) [inline digester.go:66]: Should you check if t.closed is true, and return success or an error, rather then executing it a second time?

Comment (nalind) [reply digester.go:69]: Can't hurt.  Added a check there.

**Comment by rhatdan** [comment]
LGTM
@TomSweeneyRedHat @giuseppe @vrothberg @QiWang19 @ashley-cui PTAL

File: digester.go (hunk 1 above)
Comment (nalind) [reply digester.go:72]: Good news: we're definitely running the unit tests.  Bad news: they found a race.  Good news: added a lock to fix it.

**Comment by TomSweeneyRedHat** [comment]
LGTM 
assuming happy tests

**Pull request merged**